typescript/ws-protocol/src/FoxgloveServer.ts

```diff
@@ -3,25 +3,33 @@ import { EventEmitter, EventNames, EventListener } from "eventemitter3";
 
 import { ChannelId, StatusLevel } from ".";
 import {
+  BinaryOpcode,
   Channel,
+  ClientChannel,
   ClientMessage,
-  SubscriptionId,
-  ServerMessage,
-  BinaryOpcode,
+  ClientPublish,
   IWebSocket,
+  ServerCapability,
+  ServerMessage,
+  SubscriptionId,
 } from "./types";
 
 type EventTypes = {
+  error: (error: Error) => void;
+
   subscribe: (channel: ChannelId) => void;
   unsubscribe: (channel: ChannelId) => void;
-  error: (error: Error) => void;
+  message: (event: ClientPublish) => void;
+  advertise: (channel: ClientChannel) => void;
+  unadvertise: (channel: ChannelId) => void;
 };
 
 type ClientInfo = {
   name: string;
   connection: IWebSocket;
   subscriptions: Map<SubscriptionId, ChannelId>;
   subscriptionsByChannel: Map<ChannelId, Set<SubscriptionId>>;
+  advertisements: Map<ChannelId, ClientChannel>;
```

Comment:
nit: maybe make a separate type for ClientChannelId since it is a distinct namespace from server channel ids? (Ideally we should use tagged/branded types to actually prevent misuse...)